README.md
# yassp-server

yassp-server was **a part of** an internal closed source project called Yet Another ShadowSocks Panel (YaSSP). Like [ss-panel](https://github.com/orvice/ss-panel/), YaSSP was a full-functional Shadowsocks panel with user management, accounting, sales, et al. However, this project has been dead.

```
  +---------+            +---------------+-------------+
  |         |    HTTP    |               |             |
  |  YaSSP  | <--------> | yassp-server <-> SS servers |
  |         |            |  (this repo)  |             |
  +---------+            +---------------+-------------+
single control              may have multiple instances
    server
```

yassp-server is a daemon progrom running on server where SS servers run. It synchronizes user configurations and traffics with YaSSP, and is responsible for run/stop/restart SS servers.

Since the original project has been **dead**, this repo is **unlikely to be maintained**.

## Features
* Support both SS-Python and SS-libev ports (see [ssmanager](https://github.com/sorz/ssmanager/))
* Pull SS configs via HTTP API
* Also an optional push API for update SS configs
* Configurable hostname & pre-shared key pair for multiple servers
* Accounting: push traffic usage via HTTP API
* Traffic usage won't lost on temporary network/server issues and normal shutdown/restart


## API

Based on JSON & HTTP. Variables are configured on `config.cfg`.


### Profiles
A JSON list of SS configs. For example,
```json
[{"port": 8001, "method": "aes-256-gcm", "password": "password"},
 {"port": 8002, "method": "chacha20-ietf-poly1305", "password": "test123"}]
```
All arguments are passed to `ssmanager` and then to SS server eventually, see docs of SS for available args.

### Pull configs
yassp-server will pull SS configs from `{yassp url}`.

* HTTP `GET` `{yassp url}/services/`
* With HTTP basic auth `{yassp hostname}:{yassp psk}`
* Return `200 OK`, `Profiles` as body (see previous section), or 
* `304 Not Modified` if configs not change since last pull *.

\* Both `last_modified` and `etag` headers are supported.

### Push configs

If enabled, a simple HTTP server will run on `{push bind address}:{push bind port}`.

To push configs, send a `POST` to `/instances?token={push token}`,
with `Profiles` as body. `204 No Content` will returned.

### Traffic stats

Traffic stats will be `PUSH`ed to `{yassp url}/traffics/`. Body is a JSON dict of `{port number: traffic in bytes since last upload}`, Auth header is the same as pull configs'.



debian/changelog
yassp-server (0.1.0-4) unstable; urgency=medium

  * Multiple sites sharing one config

 --  <xierch@zh1.li>  Tue, 20 Dec 2016 16:59:02 +0800

yassp-server (0.1.0-2) unstable; urgency=medium

  * Change systemd file (restart on failure)

 --  <xierch@zh1.li>  Tue, 20 Dec 2016 16:05:46 +0800

yassp-server (0.1.0-1) unstable; urgency=medium

  * Initial release

 -- Shell Chen <yassp@sorz.org>  Fri, 09 Dec 2016 18:59:13 +0000


yasspserver/__main__.py
#!/usr/bin/env python3
import os
import sys
import time
import signal
import logging
from os.path import isfile
from configparser import ConfigParser
from ssmanager import Server
from ssmanager.sslibev import Manager as ManagerEv
from ssmanager.sspy import Manager as ManagerPy

from .yassp import YaSSP
from . import pushserver, utils


def get_config():
    if len(sys.argv) not in (2, 3):
        print('Usage:\n\t%s <CONFIG-FILE> [section-name]' % sys.argv[0],
              file=sys.stderr)
        sys.exit(1)
    path = sys.argv[1]
    section = sys.argv[2] if len(sys.argv) > 2 else 'DEFAULT'
    if not isfile(path):
        print('Error: config file "%s" not exist.' % path, file=sys.stderr)
        sys.exit(1)
    config = ConfigParser()
    config.read(path)
    try:
        return config[section]
    except KeyError as e:
        print('Error: section [%s] not found in config file.' % section, file=sys.stderr)
        sys.exit(1)


def exit(signalnum, frame):
    logging.info('Stopping by SIGTERM/SIGHUP.')
    sys.exit()

def main():
    conf = get_config()
    logging.basicConfig(level=conf.getint('log level'),
                        format='%(asctime)s %(levelname)-s: %(message)s')
    utils.ss_bind_address = conf['ss-server bind']
    ss_type = conf['ss daemon type']
    if ss_type == 'libev':
        Manager = ManagerEv
    elif ss_type == 'python':
        Manager = ManagerPy
    else:
        logging.critical('Unknown "ss daemon type" %s, must be "libev" or "python"', ss_type)
        sys.exit(1)
    manager = Manager(ss_bin=conf['ss-server path'],
                      print_ss_log=conf.getboolean('ss-server print log'))

    yassp = YaSSP(conf['yassp url'], conf['yassp hostname'],
                  conf['yassp psk'], manager)
    yassp.traffic_sync_threshold = conf.getint('traffic sync threshold')
    yassp.traffic_sync_timeout = conf.getint('traffic sync timeout')
    signal.signal(signal.SIGTERM, exit)
    signal.signal(signal.SIGHUP, exit)

    try:
        manager.start()
        yassp.start()
        if conf.getboolean('push server enable'):
            pushserver.run(manager, conf['push token'],
                           host=conf['push bind address'],
                           port=conf.getint('push bind port'))
        else:
            yassp._listen_thread.join()
    except KeyboardInterrupt:
        logging.info('Stopping by ^C.')
    finally:
        yassp.stop()
        manager.stop()
        logging.info('Exited.')

if __name__ == '__main__':
    main()



yasspserver/pushserver.py
from bottle import get, post, request, abort, response
from io import TextIOWrapper
import bottle
import json
import logging

from .utils import parse_servers


token = 'test123'

def _check_token():
    if request.query.token != token:
        abort(401, 'No token given or invaild token.')

@get('/')
def home():
    _check_token()
    return 'It works.'

@post('/instances')
def update_instances():
    _check_token()
    profiles = json.load(TextIOWrapper(request.body))
    logging.debug('Syncing %s profiles (push)...' % len(profiles))
    manager.update(parse_servers(profiles))
    response.status = 204

def run(ssmanager, key, *args, **kwargs):
    global token, manager
    token = key
    manager = ssmanager
    logging.info('Starting push server...')
    bottle.run(*args, **kwargs)

def main():
    bottle.run(host='0.0.0.0', port=8080, reloader=True)

if __name__ == '__main__':
    main()



yasspserver/yassp.py
import time
import json
import logging
import requests
from requests.exceptions import RequestException, ConnectionError
from threading import Thread
from collections import defaultdict
from urllib.parse import urljoin

from .utils import parse_servers


TRAFFIC_CHECK_PERIOD = 30

class YaSSP():
    traffic_sync_threshold = 100 * 1024 * 1024  # 100 MiB
    traffic_sync_timeout = 60 * 30  # 30 mins

    def __init__(self, url_prefix, hostname, psk, manager):
        self._running = False
        self._synced_traffic = defaultdict(lambda: 0)
        self._earliest_unsynced_time = {}
        self._http_last_modified = {}
        self._http_etag = {}

        self._url_prefix = url_prefix
        self._hostname = hostname
        self._psk = psk
        self._manager = manager

    def _request(self, func, path, **kwargs):
        auth=(self._hostname, self._psk)
        if 'timeout' not in kwargs:
            kwargs['timeout'] = 10
        headers = {}
        if path in self._http_last_modified:
            headers['if-modified-since'] = self._http_last_modified[path]
        if path in self._http_etag:
            headers['if-none-match'] = self._http_etag[path]

        resp = func(urljoin(self._url_prefix, path),
                    auth=auth, headers=headers, **kwargs)

        if 'last_modified' in resp.headers:
            self._http_last_modified[path] = resp.headers['last_modified']
        if 'etag' in resp.headers:
            self._http_etag[path] = resp.headers['etag']

        if resp.status_code == 403:
            raise AuthenticationError()
        elif resp.status_code == 200:
            return resp.json()
        elif resp.status_code in (204, 304):
            return
        else:
            raise UnexpectedResponseError()

    def _get(self, *args, **kwargs):
        return self._request(requests.get, *args, **kwargs)

    def _post(self, *args, **kwargs):
        return self._request(requests.post, *args, **kwargs)

    def start(self):
        self._listen_thread = Thread(target=self._listen_profile_changes, daemon=True)
        self._traffic_thread = Thread(target=self._traffic_timer, daemon=True)
        self._running = True
        self.update_profiles()
        self._listen_thread.start()
        self._traffic_thread.start()

    def stop(self):
        self._running = False
        self.update_traffic(force_all=True)

    def update_profiles(self):
        try:
            profiles = self._get('services/')
        except (RequestException, AuthenticationError, UnexpectedResponseError,
                ConnectionError, ValueError, KeyError) as e:
            logging.warning('Error on update profiles: %s' % e)
        else:
            if profiles is not None:
                self._manager.update(parse_servers(profiles))
                logging.debug('Syncing %s profiles (pull)...' % len(profiles))

    def update_traffic(self, force_all=False):
        stat = self._manager.stat()
        to_upload = {}
        for port, traffic in stat.items():
            increment = traffic - self._synced_traffic[port]
            if increment == 0:
                continue
            if increment < 0:
                self._synced_traffic[port] = 0
                increment = traffic

            if port not in self._earliest_unsynced_time:
                self._earliest_unsynced_time[port] = time.time()
            if time.time() - self._earliest_unsynced_time[port] > self.traffic_sync_timeout \
               or increment >= self.traffic_sync_threshold \
               or force_all:
                to_upload[port] = increment

        if to_upload:
            logging.debug('Uploading traffic (%d)...' % len(to_upload))
            logging.debug(to_upload)
            try:
                self._post('traffics/', data=json.dumps(to_upload))
            except (RequestException, AuthenticationError, UnexpectedResponseError, ConnectionError) as e:
                logging.warning('Error on upload traffic: %s' % e)
            else:
                for port, __ in to_upload.items():
                    self._synced_traffic[port] = stat[port]
                    if port in self._earliest_unsynced_time:
                        del self._earliest_unsynced_time[port]

    def _listen_profile_changes(self):
        # Currently, we just fetch all profiles every 1 minutes.
        timeout = 60 * 1
        time.sleep(timeout)
        while self._running:
            self.update_profiles()
            time.sleep(timeout)

    def _traffic_timer(self):
        while self._running:
            time.sleep(min(TRAFFIC_CHECK_PERIOD, self.traffic_sync_timeout))
            self.update_traffic()


class AuthenticationError(Exception):
    pass

class UnexpectedResponseError(Exception):
    pass

